File: docs/beginner/module-02-basic-search/code-samples/javascript/04_wildcard_search.js
/**
 * Wildcard Search - Module 2 JavaScript Examples
 * Pattern matching with wildcards in Azure AI Search using JavaScript SDK
 * 
 * This module demonstrates:
 * - Prefix matching with *
 * - Suffix matching with *
 * - Pattern matching strategies
 * - When to use wildcards
 * - Wildcard search limitations
 */

const { SearchClient, AzureKeyCredential } = require('@azure/search-documents');

class WildcardSearch {
    /**
     * Initialize with a search client
     * @param {SearchClient} searchClient - Azure Search client
     */
    constructor(searchClient) {
        if (!searchClient) {
            throw new Error('SearchClient is required');
        }
        this.searchClient = searchClient;
    }

    /**
     * Search for terms starting with a prefix (prefix*)
     * @param {string} prefix - The prefix to search for
     * @param {number} top - Maximum number of results to return
     * @returns {Promise<Object>} Search results
     */
    async prefixSearch(prefix, top = 10) {
        try {
            const query = `${prefix}*`;
            console.log(`Performing prefix search: '${query}'`);

            const searchOptions = {
                top: top,
                includeTotalCount: true
            };

            const results = await this.searchClient.search(query, searchOptions);
            
            const resultArray = [];
            for await (const result of results.results) {
                resultArray.push(result);
            }

            console.log(`Found ${resultArray.length} results with prefix '${prefix}'`);
            return {
                results: resultArray,
                totalCount: results.count
            };
        } catch (error) {
            console.error(`Error in prefix search: ${error.message}`);
            return { results: [], totalCount: 0 };
        }
    }

    /**
     * Search for terms ending with a suffix (*suffix)
     * @param {string} suffix - The suffix to search for
     * @param {number} top - Maximum number of results to return
     * @returns {Promise<Object>} Search results
     */
    async suffixSearch(suffix, top = 10) {
        try {
            const query = `*${suffix}`;
            console.log(`Performing suffix search: '${query}'`);

            const searchOptions = {
                top: top,
                includeTotalCount: true
            };

            const results = await this.searchClient.search(query, searchOptions);
            
            const resultArray = [];
            for await (const result of results.results) {
                resultArray.push(result);
            }

            console.log(`Found ${resultArray.length} results with suffix '${suffix}'`);
            return {
                results: resultArray,
                totalCount: results.count
            };
        } catch (error) {
            console.error(`Error in suffix search: ${error.message}`);
            return { results: [], totalCount: 0 };
        }
    }

    /**
     * Search for terms containing a substring (*substring*)
     * @param {string} substring - The substring to search for
     * @param {number} top - Maximum number of results to return
     * @returns {Promise<Object>} Search results
     */
    async containsSearch(substring, top = 10) {
        try {
            const query = `*${substring}*`;
            console.log(`Performing contains search: '${query}'`);

            const searchOptions = {
                top: top,
                includeTotalCount: true
            };

            const results = await this.searchClient.search(query, searchOptions);
            
            const resultArray = [];
            for await (const result of results.results) {
                resultArray.push(result);
            }

            console.log(`Found ${resultArray.length} results containing '${substring}'`);
            return {
                results: resultArray,
                totalCount: results.count
            };
        } catch (error) {
            console.error(`Error in contains search: ${error.message}`);
            return { results: [], totalCount: 0 };
        }
    }

    /**
     * Search for multiple wildcard patterns
     * @param {Array<string>} patterns - List of wildcard patterns to search for
     * @param {number} top - Maximum results per pattern
     * @returns {Promise<Object>} Object mapping patterns to their results
     */
    async multipleWildcardSearch(patterns, top = 5) {
        const results = {};

        for (const pattern of patterns) {
            try {
                console.log(`Searching for pattern: '${pattern}'`);

                const searchOptions = {
                    top: top,
                    includeTotalCount: true
                };

                const searchResults = await this.searchClient.search(pattern, searchOptions);
                
                const resultArray = [];
                for await (const result of searchResults.results) {
                    resultArray.push(result);
                }

                results[pattern] = {
                    results: resultArray,
                    totalCount: searchResults.count
                };
            } catch (error) {
                console.error(`Error searching pattern '${pattern}': ${error.message}`);
                results[pattern] = { results: [], totalCount: 0 };
            }
        }

        return results;
    }

    /**
     * Compare different wildcard patterns for the same base term
     * @param {string} baseTerm - Base term to create patterns from
     * @param {number} top - Maximum results per pattern
     * @returns {Promise<Object>} Object mapping pattern types to their results
     */
    async compareWildcardPatterns(baseTerm, top = 5) {
        const patterns = {
            'exact': baseTerm,
            'prefix': `${baseTerm}*`,
            'suffix': `*${baseTerm}`,
            'contains': `*${baseTerm}*`
        };

        const results = {};

        for (const [patternType, pattern] of Object.entries(patterns)) {
            try {
                const searchOptions = {
                    top: top,
                    includeTotalCount: true
                };

                const searchResults = await this.searchClient.search(pattern, searchOptions);
                
                const resultArray = [];
                for await (const result of searchResults.results) {
                    resultArray.push(result);
                }

                results[patternType] = {
                    results: resultArray,
                    totalCount: searchResults.count
                };
            } catch (error) {
                console.error(`Error in ${patternType} search: ${error.message}`);
                results[patternType] = { results: [], totalCount: 0 };
            }
        }

        return results;
    }

    /**
     * Display comparison of wildcard patterns
     * @param {string} baseTerm - Base term used for patterns
     * @param {Object} results - Results from different wildcard patterns
     */
    static displayWildcardComparison(baseTerm, results) {
        console.log(`\n🃏 Wildcard Patterns Comparison: '${baseTerm}'`);
        console.log('='.repeat(60));

        const patternInfo = {
            'exact': ['Exact Match', baseTerm, 'Exact term only'],
            'prefix': ['Prefix Match', `${baseTerm}*`, `Terms starting with "${baseTerm}"`],
            'suffix': ['Suffix Match', `*${baseTerm}`, `Terms ending with "${baseTerm}"`],
            'contains': ['Contains Match', `*${baseTerm}*`, `Terms containing "${baseTerm}"`]
        };

        for (const [patternType, patternResults] of Object.entries(results)) {
            if (patternInfo[patternType]) {
                const [name, pattern, description] = patternInfo[patternType];

                console.log(`\n${name}:`);
                console.log(`   Pattern: ${pattern}`);
                console.log(`   Description: ${description}`);
                console.log(`   Results found: ${patternResults.results.length}`);

                if (patternResults.results.length > 0) {
                    console.log('   Top matches:');
                    patternResults.results.slice(0, 3).forEach((result, index) => {
                        const title = result.document.title || 'No title';
                        const score = result.score || 0.0;
                        console.log(`     ${index + 1}. ${title} (Score: ${score.toFixed(3)})`);
                    });
                } else {
                    console.log('   No matches found');
                }

                console.log('-'.repeat(40));
            }
        }

        // Analysis
        console.log(`\n📊 PATTERN ANALYSIS:`);
        const exactCount = results.exact?.results.length || 0;
        const prefixCount = results.prefix?.results.length || 0;
        const suffixCount = results.suffix?.results.length || 0;
        const containsCount = results.contains?.results.length || 0;

        console.log(`   Exact: ${exactCount} (most specific)`);
        console.log(`   Prefix: ${prefixCount}`);
        console.log(`   Suffix: ${suffixCount}`);
        console.log(`   Contains: ${containsCount} (broadest)`);

        if (containsCount >= Math.max(prefixCount, suffixCount) && 
            Math.max(prefixCount, suffixCount) >= exactCount) {
            console.log('   ✅ Expected pattern: Contains ≥ Prefix/Suffix ≥ Exact');
        } else {
            console.log('   ⚠️ Unexpected pattern - depends on your data');
        }
    }
}

/**
 * Demonstrate wildcard search operations
 */
async function demonstrateWildcardSearch() {
    console.log('🃏 Wildcard Search Demonstration');
    console.log('='.repeat(50));

    try {
        // Initialize search client (replace with your actual service details)
        const searchClient = new SearchClient(
            'https://your-service.search.windows.net',
            'your-index-name',
            new AzureKeyCredential('your-api-key')
        );

        const searchOps = new WildcardSearch(searchClient);

        // Example 1: Compare wildcard patterns
        console.log('\n1️⃣ Wildcard Patterns Comparison');

        const baseTerm = 'program';
        const wildcardResults = await searchOps.compareWildcardPatterns(baseTerm, 5);
        WildcardSearch.displayWildcardComparison(baseTerm, wildcardResults);

        // Example 2: Practical wildcard searches
        console.log('\n' + '='.repeat(70));
        console.log('\n2️⃣ Practical Wildcard Examples');
        console.log('-'.repeat(30));

        const practicalExamples = [
            { description: 'Find programming languages', pattern: 'program*', type: 'prefix' },
            { description: 'Find development terms', pattern: '*develop*', type: 'contains' },
            { description: 'Find tutorial content', pattern: '*tutorial', type: 'suffix' },
            { description: 'Find JavaScript variations', pattern: 'java*', type: 'prefix' }
        ];

        for (const example of practicalExamples) {
            console.log(`\n📋 ${example.description}`);
            console.log(`   Pattern: ${example.pattern} (${example.type})`);

            let results;
            switch (example.type) {
                case 'prefix':
                    results = await searchOps.prefixSearch(example.pattern.slice(0, -1), 3);
                    break;
                case 'suffix':
                    results = await searchOps.suffixSearch(example.pattern.slice(1), 3);
                    break;
                case 'contains':
                    results = await searchOps.containsSearch(example.pattern.slice(1, -1), 3);
                    break;
            }

            console.log(`   Results: ${results.results.length} found`);

            if (results.results.length > 0) {
                const topResult = results.results[0];
                const title = topResult.document.title || 'No title';
                const score = topResult.score || 0.0;
                console.log(`   Top match: ${title} (Score: ${score.toFixed(3)})`);
            }
        }

        // Example 3: Multiple wildcard search
        console.log('\n' + '='.repeat(70));
        console.log('\n3️⃣ Multiple Wildcard Search');
        console.log('-'.repeat(30));

        const wildcardPatterns = [
            'web*',      // web, website, webdev, etc.
            '*script',   // javascript, typescript, etc.
            '*data*',    // database, metadata, etc.
            'api*'       // api, apis, etc.
        ];

        const multiResults = await searchOps.multipleWildcardSearch(wildcardPatterns, 2);

        for (const [pattern, results] of Object.entries(multiResults)) {
            console.log(`\nPattern: ${pattern}`);
            console.log(`Results: ${results.results.length}`);
            
            if (results.results.length > 0) {
                results.results.forEach((result, index) => {
                    const title = result.document.title || 'No title';
                    const score = result.score || 0.0;
                    console.log(`  ${index + 1}. ${title} (Score: ${score.toFixed(3)})`);
                });
            }
        }

        console.log('\n✅ Wildcard search demonstration completed!');

    } catch (error) {
        console.error(`❌ Demo failed: ${error.message}`);
        console.log('Make sure your Azure AI Search service is configured correctly.');
    }
}

/**
 * Demonstrate best practices for wildcard search
 */
async function wildcardSearchBestPractices() {
    console.log('\n📚 Wildcard Search Best Practices');
    console.log('='.repeat(50));

    console.log('\n💡 When to Use Wildcards:');
    console.log('\n✅ Prefix Search (term*):');
    console.log('   - Finding word variations (program, programming, programmer)');
    console.log('   - Technology families (java, javascript, javadoc)');
    console.log('   - Brand or product lines (micro, microsoft, microservice)');
    console.log('   - Language variations (develop, developer, development)');

    console.log('\n✅ Suffix Search (*term):');
    console.log('   - Finding words with common endings (*ing, *tion, *ment)');
    console.log('   - File types or extensions (*script, *doc)');
    console.log('   - Categories or types (*tutorial, *guide)');

    console.log('\n✅ Contains Search (*term*):');
    console.log('   - Finding partial matches when unsure of exact form');
    console.log('   - Searching within compound words');
    console.log('   - When you know a key part but not the whole term');

    console.log('\n⚠️ Wildcard Limitations:');
    console.log('   ❌ Can be slower than exact searches');
    console.log('   ❌ May return too many irrelevant results');
    console.log('   ❌ Leading wildcards (*term) are generally slower');
    console.log('   ❌ Multiple wildcards in one term can be very slow');

    console.log('\n🔧 Performance Tips:');
    console.log('   ✅ Use specific prefixes (at least 2-3 characters)');
    console.log('   ✅ Combine with other terms to narrow results');
    console.log('   ✅ Prefer prefix wildcards over suffix wildcards');
    console.log('   ✅ Test wildcard queries with small result sets first');

    // Demonstrate good vs bad practices
    console.log('\n🧪 Good vs Bad Wildcard Examples:');

    console.log('\n✅ Good Wildcard Practices:');
    const goodExamples = [
        ['program*', 'Specific prefix, likely to find relevant terms'],
        ['java* AND tutorial', 'Wildcard combined with specific term'],
        ['*development', 'Common suffix, reasonable scope']
    ];

    goodExamples.forEach(([pattern, explanation]) => {
        console.log(`   ${pattern}: ${explanation}`);
    });

    console.log('\n❌ Problematic Wildcard Practices:');
    const badExamples = [
        ['*a*', 'Too broad, will match almost everything'],
        ['*', 'Matches all documents, not useful for search'],
        ['*e*t*', 'Multiple wildcards, very slow']
    ];

    badExamples.forEach(([pattern, explanation]) => {
        console.log(`   ${pattern}: ${explanation}`);
    });

    // Show optimization example
    console.log('\n🚀 Optimization Example:');
    console.log('   Instead of: \'*script\'');
    console.log('   Try: \'javascript OR typescript OR script\'');
    console.log('   Benefit: More specific, faster, better relevance');
}

// Main execution
async function main() {
    try {
        await demonstrateWildcardSearch();
        await wildcardSearchBestPractices();

        console.log('\n💡 Next Steps:');
        console.log('   - Experiment with different wildcard patterns');
        console.log('   - Try combining wildcards with boolean operators');
        console.log('   - Check out 05_field_search.js for field-specific searches');
        console.log('   - Learn about search parameters in other examples');
    } catch (error) {
        console.error(`❌ Main execution failed: ${error.message}`);
    }
}

// Export for use in other modules
module.exports = {
    WildcardSearch,
    demonstrateWildcardSearch,
    wildcardSearchBestPractices
};

// Run if this file is executed directly
if (require.main === module) {
    main();
}

File: docs/beginner/module-02-basic-search/code-samples/csharp/04_WildcardSearch.cs
/*
Wildcard Search - Module 2 C# Examples
Pattern matching with wildcards in Azure AI Search using .NET SDK

This module demonstrates:
- Prefix matching with *
- Suffix matching with *
- Pattern matching strategies
- When to use wildcards
- Wildcard search limitations
*/

using System;
using System.Collections.Generic;
using System.Linq;
using System.Threading.Tasks;
using Azure;
using Azure.Search.Documents;
using Azure.Search.Documents.Models;

namespace AzureSearchHandbook.Module02.BasicSearch
{
    public class WildcardSearch
    {
        private readonly SearchClient _searchClient;

        public WildcardSearch(SearchClient searchClient)
        {
            _searchClient = searchClient ?? throw new ArgumentNullException(nameof(searchClient));
        }

        /// <summary>
        /// Search for terms starting with a prefix (prefix*)
        /// </summary>
        /// <param name="prefix">The prefix to search for</param>
        /// <param name="top">Maximum number of results to return</param>
        /// <returns>Search results</returns>
        public async Task<SearchResults<SearchDocument>> PrefixSearchAsync(string prefix, int top = 10)
        {
            try
            {
                var query = $"{prefix}*";
                Console.WriteLine($"Performing prefix search: '{query}'");

                var searchOptions = new SearchOptions
                {
                    Size = top,
                    IncludeTotalCount = true
                };

                var results = await _searchClient.SearchAsync<SearchDocument>(query, searchOptions);
                
                Console.WriteLine($"Found results with prefix '{prefix}'");
                return results.Value;
            }
            catch (Exception ex)
            {
                Console.WriteLine($"Error in prefix search: {ex.Message}");
                throw;
            }
        }

        /// <summary>
        /// Search for terms ending with a suffix (*suffix)
        /// </summary>
        /// <param name="suffix">The suffix to search for</param>
        /// <param name="top">Maximum number of results to return</param>
        /// <returns>Search results</returns>
        public async Task<SearchResults<SearchDocument>> SuffixSearchAsync(string suffix, int top = 10)
        {
            try
            {
                var query = $"*{suffix}";
                Console.WriteLine($"Performing suffix search: '{query}'");

                var searchOptions = new SearchOptions
                {
                    Size = top,
                    IncludeTotalCount = true
                };

                var results = await _searchClient.SearchAsync<SearchDocument>(query, searchOptions);
                
                Console.WriteLine($"Found results with suffix '{suffix}'");
                return results.Value;
            }
            catch (Exception ex)
            {
                Console.WriteLine($"Error in suffix search: {ex.Message}");
                throw;
            }
        }

        /// <summary>
        /// Search for terms containing a substring (*substring*)
        /// </summary>
        /// <param name="substring">The substring to search for</param>
        /// <param name="top">Maximum number of results to return</param>
        /// <returns>Search results</returns>
        public async Task<SearchResults<SearchDocument>> ContainsSearchAsync(string substring, int top = 10)
        {
            try
            {
                var query = $"*{substring}*";
                Console.WriteLine($"Performing contains search: '{query}'");

                var searchOptions = new SearchOptions
                {
                    Size = top,
                    IncludeTotalCount = true
                };

                var results = await _searchClient.SearchAsync<SearchDocument>(query, searchOptions);
                
                Console.WriteLine($"Found results containing '{substring}'");
                return results.Value;
            }
            catch (Exception ex)
            {
                Console.WriteLine($"Error in contains search: {ex.Message}");
                throw;
            }
        }

        /// <summary>
        /// Search for multiple wildcard patterns
        /// </summary>
        /// <param name="patterns">List of wildcard patterns to search for</param>
        /// <param name="top">Maximum results per pattern</param>
        /// <returns>Dictionary mapping patterns to their results</returns>
        public async Task<Dictionary<string, SearchResults<SearchDocument>>> MultipleWildcardSearchAsync(
            IEnumerable<string> patterns, int top = 5)
        {
            var results = new Dictionary<string, SearchResults<SearchDocument>>();

            foreach (var pattern in patterns)
            {
                try
                {
                    Console.WriteLine($"Searching for pattern: '{pattern}'");

                    var searchOptions = new SearchOptions
                    {
                        Size = top,
                        IncludeTotalCount = true
                    };

                    var searchResults = await _searchClient.SearchAsync<SearchDocument>(pattern, searchOptions);
                    results[pattern] = searchResults.Value;
                }
                catch (Exception ex)
                {
                    Console.WriteLine($"Error searching pattern '{pattern}': {ex.Message}");
                    // Create empty result for failed patterns
                    results[pattern] = null;
                }
            }

            return results;
        }

        /// <summary>
        /// Compare different wildcard patterns for the same base term
        /// </summary>
        /// <param name="baseTerm">Base term to create patterns from</param>
        /// <param name="top">Maximum results per pattern</param>
        /// <returns>Dictionary mapping pattern types to their results</returns>
        public async Task<Dictionary<string, SearchResults<SearchDocument>>> CompareWildcardPatternsAsync(
            string baseTerm, int top = 5)
        {
            var patterns = new Dictionary<string, string>
            {
                ["exact"] = baseTerm,
                ["prefix"] = $"{baseTerm}*",
                ["suffix"] = $"*{baseTerm}",
                ["contains"] = $"*{baseTerm}*"
            };

            var results = new Dictionary<string, SearchResults<SearchDocument>>();

            foreach (var kvp in patterns)
            {
                var patternType = kvp.Key;
                var pattern = kvp.Value;

                try
                {
                    var searchOptions = new SearchOptions
                    {
                        Size = top,
                        IncludeTotalCount = true
                    };

                    var searchResults = await _searchClient.SearchAsync<SearchDocument>(pattern, searchOptions);
                    results[patternType] = searchResults.Value;
                }
                catch (Exception ex)
                {
                    Console.WriteLine($"Error in {patternType} search: {ex.Message}");
                    results[patternType] = null;
                }
            }

            return results;
        }

        /// <summary>
        /// Display comparison of wildcard patterns
        /// </summary>
        /// <param name="baseTerm">Base term used for patterns</param>
        /// <param name="results">Results from different wildcard patterns</param>
        public static void DisplayWildcardComparison(string baseTerm, 
            Dictionary<string, SearchResults<SearchDocument>> results)
        {
            Console.WriteLine($"\n🃏 Wildcard Patterns Comparison: '{baseTerm}'");
            Console.WriteLine(new string('=', 60));

            var patternInfo = new Dictionary<string, (string Name, string Pattern, string Description)>
            {
                ["exact"] = ("Exact Match", baseTerm, "Exact term only"),
                ["prefix"] = ("Prefix Match", $"{baseTerm}*", $"Terms starting with \"{baseTerm}\""),
                ["suffix"] = ("Suffix Match", $"*{baseTerm}", $"Terms ending with \"{baseTerm}\""),
                ["contains"] = ("Contains Match", $"*{baseTerm}*", $"Terms containing \"{baseTerm}\"")
            };

            foreach (var kvp in results)
            {
                var patternType = kvp.Key;
                var patternResults = kvp.Value;

                if (patternInfo.ContainsKey(patternType))
                {
                    var (name, pattern, description) = patternInfo[patternType];
                    var resultList = patternResults?.GetResults()?.ToList() ?? new List<SearchResult<SearchDocument>>();

                    Console.WriteLine($"\n{name}:");
                    Console.WriteLine($"   Pattern: {pattern}");
                    Console.WriteLine($"   Description: {description}");
                    Console.WriteLine($"   Results found: {resultList.Count}");

                    if (resultList.Any())
                    {
                        Console.WriteLine("   Top matches:");
                        for (int i = 0; i < Math.Min(3, resultList.Count); i++)
                        {
                            var result = resultList[i];
                            var title = result.Document.TryGetValue("title", out var titleValue) ? 
                                titleValue?.ToString() : "No title";
                            var score = result.Score ?? 0.0;
                            Console.WriteLine($"     {i + 1}. {title} (Score: {score:F3})");
                        }
                    }
                    else
                    {
                        Console.WriteLine("   No matches found");
                    }

                    Console.WriteLine(new string('-', 40));
                }
            }

            // Analysis
            Console.WriteLine($"\n📊 PATTERN ANALYSIS:");
            var exactCount = results.ContainsKey("exact") && results["exact"] != null ? 
                results["exact"].GetResults().Count() : 0;
            var prefixCount = results.ContainsKey("prefix") && results["prefix"] != null ? 
                results["prefix"].GetResults().Count() : 0;
            var suffixCount = results.ContainsKey("suffix") && results["suffix"] != null ? 
                results["suffix"].GetResults().Count() : 0;
            var containsCount = results.ContainsKey("contains") && results["contains"] != null ? 
                results["contains"].GetResults().Count() : 0;

            Console.WriteLine($"   Exact: {exactCount} (most specific)");
            Console.WriteLine($"   Prefix: {prefixCount}");
            Console.WriteLine($"   Suffix: {suffixCount}");
            Console.WriteLine($"   Contains: {containsCount} (broadest)");

            if (containsCount >= Math.Max(prefixCount, suffixCount) && 
                Math.Max(prefixCount, suffixCount) >= exactCount)
            {
                Console.WriteLine("   ✅ Expected pattern: Contains ≥ Prefix/Suffix ≥ Exact");
            }
            else
            {
                Console.WriteLine("   ⚠️ Unexpected pattern - depends on your data");
            }
        }
    }

    /// <summary>
    /// Program class for demonstration
    /// </summary>
    public class WildcardSearchProgram
    {
        // Replace with your actual service details
        private const string ServiceEndpoint = "https://your-service.search.windows.net";
        private const string ApiKey = "your-api-key";
        private const string IndexName = "your-index-name";

        public static async Task Main(string[] args)
        {
            try
            {
                // Initialize search client
                var searchClient = new SearchClient(
                    new Uri(ServiceEndpoint),
                    IndexName,
                    new AzureKeyCredential(ApiKey)
                );

                await DemonstrateWildcardSearchAsync(searchClient);
                await WildcardSearchBestPracticesAsync(searchClient);

                Console.WriteLine("\n💡 Next Steps:");
                Console.WriteLine("   - Experiment with different wildcard patterns");
                Console.WriteLine("   - Try combining wildcards with boolean operators");
                Console.WriteLine("   - Check out 05_FieldSearch.cs for field-specific searches");
                Console.WriteLine("   - Learn about search parameters in other examples");
            }
            catch (Exception ex)
            {
                Console.WriteLine($"❌ Demo failed: {ex.Message}");
                Console.WriteLine("Make sure your Azure AI Search service is configured correctly.");
            }
        }

        private static async Task DemonstrateWildcardSearchAsync(SearchClient searchClient)
        {
            Console.WriteLine("🃏 Wildcard Search Demonstration");
            Console.WriteLine(new string('=', 50));

            var searchOps = new WildcardSearch(searchClient);

            // Example 1: Compare wildcard patterns
            Console.WriteLine("\n1️⃣ Wildcard Patterns Comparison");

            var baseTerm = "program";
            var wildcardResults = await searchOps.CompareWildcardPatternsAsync(baseTerm, 5);
            WildcardSearch.DisplayWildcardComparison(baseTerm, wildcardResults);

            // Example 2: Practical wildcard searches
            Console.WriteLine($"\n{new string('=', 70)}");
            Console.WriteLine("\n2️⃣ Practical Wildcard Examples");
            Console.WriteLine(new string('-', 30));

            var practicalExamples = new[]
            {
                new { Description = "Find programming languages", Pattern = "program*", Type = "prefix" },
                new { Description = "Find development terms", Pattern = "*develop*", Type = "contains" },
                new { Description = "Find tutorial content", Pattern = "*tutorial", Type = "suffix" },
                new { Description = "Find JavaScript variations", Pattern = "java*", Type = "prefix" }
            };

            foreach (var example in practicalExamples)
            {
                Console.WriteLine($"\n📋 {example.Description}");
                Console.WriteLine($"   Pattern: {example.Pattern} ({example.Type})");

                SearchResults<SearchDocument> results = null;
                switch (example.Type)
                {
                    case "prefix":
                        results = await searchOps.PrefixSearchAsync(example.Pattern.TrimEnd('*'), 3);
                        break;
                    case "suffix":
                        results = await searchOps.SuffixSearchAsync(example.Pattern.TrimStart('*'), 3);
                        break;
                    case "contains":
                        results = await searchOps.ContainsSearchAsync(example.Pattern.Trim('*'), 3);
                        break;
                }

                if (results != null)
                {
                    var resultList = results.GetResults().ToList();
                    Console.WriteLine($"   Results: {resultList.Count} found");

                    if (resultList.Any())
                    {
                        var topResult = resultList.First();
                        var title = topResult.Document.TryGetValue("title", out var titleValue) ? 
                            titleValue?.ToString() : "No title";
                        var score = topResult.Score ?? 0.0;
                        Console.WriteLine($"   Top match: {title} (Score: {score:F3})");
                    }
                }
            }

            // Example 3: Multiple wildcard search
            Console.WriteLine($"\n{new string('=', 70)}");
            Console.WriteLine("\n3️⃣ Multiple Wildcard Search");
            Console.WriteLine(new string('-', 30));

            var wildcardPatterns = new[]
            {
                "web*",      // web, website, webdev, etc.
                "*script",   // javascript, typescript, etc.
                "*data*",    // database, metadata, etc.
                "api*"       // api, apis, etc.
            };

            var multiResults = await searchOps.MultipleWildcardSearchAsync(wildcardPatterns, 2);

            foreach (var kvp in multiResults)
            {
                var pattern = kvp.Key;
                var results = kvp.Value;
                var resultList = results?.GetResults()?.ToList() ?? new List<SearchResult<SearchDocument>>();

                Console.WriteLine($"\nPattern: {pattern}");
                Console.WriteLine($"Results: {resultList.Count}");
                
                if (resultList.Any())
                {
                    for (int i = 0; i < resultList.Count; i++)
                    {
                        var result = resultList[i];
                        var title = result.Document.TryGetValue("title", out var titleValue) ? 
                            titleValue?.ToString() : "No title";
                        var score = result.Score ?? 0.0;
                        Console.WriteLine($"  {i + 1}. {title} (Score: {score:F3})");
                    }
                }
            }

            Console.WriteLine("\n✅ Wildcard search demonstration completed!");
        }

        private static async Task WildcardSearchBestPracticesAsync(SearchClient searchClient)
        {
            Console.WriteLine("\n📚 Wildcard Search Best Practices");
            Console.WriteLine(new string('=', 50));

            Console.WriteLine("\n💡 When to Use Wildcards:");
            Console.WriteLine("\n✅ Prefix Search (term*):");
            Console.WriteLine("   - Finding word variations (program, programming, programmer)");
            Console.WriteLine("   - Technology families (java, javascript, javadoc)");
            Console.WriteLine("   - Brand or product lines (micro, microsoft, microservice)");
            Console.WriteLine("   - Language variations (develop, developer, development)");

            Console.WriteLine("\n✅ Suffix Search (*term):");
            Console.WriteLine("   - Finding words with common endings (*ing, *tion, *ment)");
            Console.WriteLine("   - File types or extensions (*script, *doc)");
            Console.WriteLine("   - Categories or types (*tutorial, *guide)");

            Console.WriteLine("\n✅ Contains Search (*term*):");
            Console.WriteLine("   - Finding partial matches when unsure of exact form");
            Console.WriteLine("   - Searching within compound words");
            Console.WriteLine("   - When you know a key part but not the whole term");

            Console.WriteLine("\n⚠️ Wildcard Limitations:");
            Console.WriteLine("   ❌ Can be slower than exact searches");
            Console.WriteLine("   ❌ May return too many irrelevant results");
            Console.WriteLine("   ❌ Leading wildcards (*term) are generally slower");
            Console.WriteLine("   ❌ Multiple wildcards in one term can be very slow");

            Console.WriteLine("\n🔧 Performance Tips:");
            Console.WriteLine("   ✅ Use specific prefixes (at least 2-3 characters)");
            Console.WriteLine("   ✅ Combine with other terms to narrow results");
            Console.WriteLine("   ✅ Prefer prefix wildcards over suffix wildcards");
            Console.WriteLine("   ✅ Test wildcard queries with small result sets first");

            // Demonstrate good vs bad practices
            Console.WriteLine("\n🧪 Good vs Bad Wildcard Examples:");
            var searchOps = new WildcardSearch(searchClient);

            Console.WriteLine("\n✅ Good Wildcard Practices:");
            var goodExamples = new[]
            {
                ("program*", "Specific prefix, likely to find relevant terms"),
                ("java* AND tutorial", "Wildcard combined with specific term"),
                ("*development", "Common suffix, reasonable scope")
            };

            foreach (var (pattern, explanation) in goodExamples)
            {
                Console.WriteLine($"   {pattern}: {explanation}");
            }

            Console.WriteLine("\n❌ Problematic Wildcard Practices:");
            var badExamples = new[]
            {
                ("*a*", "Too broad, will match almost everything"),
                ("*", "Matches all documents, not useful for search"),
                ("*e*t*", "Multiple wildcards, very slow")
            };

            foreach (var (pattern, explanation) in badExamples)
            {
                Console.WriteLine($"   {pattern}: {explanation}");
            }

            // Show optimization example
            Console.WriteLine("\n🚀 Optimization Example:");
            Console.WriteLine("   Instead of: '*script'");
            Console.WriteLine("   Try: 'javascript OR typescript OR script'");
            Console.WriteLine("   Benefit: More specific, faster, better relevance");
        }
    }
}